### PR TITLE
[OSDOCS#3671]: Gathering and interpreting cv history

### PR DESCRIPTION
**Pull request by obrown1205**
Version(s):
4.14+

Issue:
https://issues.redhat.com/browse/OSDOCS-3671

Link to docs preview:
https://69380--ocpdocs-pr.netlify.app/openshift-enterprise/latest/updating/troubleshooting_updates/gathering-data-cluster-update

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/gathering-clusterversion-history.adoc [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+[NOTE]
+====
+The ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the least relevant entries in the history are pruned to accommodate the limit. 
+====
+
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have installed the OpenShift CLI (`oc`).
```
Comment (petr-muller) [inline modules/gathering-clusterversion-history.adoc:19]: We can also mention that web console shows this in its "Administration -> Cluster Settings" section

File: modules/gathering-clusterversion-history.adoc [hunk 2]
```diff
@@ -0,0 +1,70 @@
+// Module included in the following assemblies:
+//
+// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="gathering-clusterversion-history_{context}"]
+= Gathering ClusterVersion history 
+
+When updating your cluster, the cluster version operator (CVO) creates a `status.history` entry to record the update. To gather and understand the history of your successful or failed cluster update, it is important to retrieve the ClusterVersion history.
```
Comment (wking) [inline modules/gathering-clusterversion-history.adoc:9]: nit: "failed" sounds so final.  The CVO has `Failing`, which is "we're currently having trouble", but the CVO will never give up on trying to reconcile the target release it's been pointed at.  And the `To gather...` sentence here seems a bit tautological ("to gather the history, you need to gather the history"), without explaining how gathering the history will help accomplish something else.  Maybe rephrase to something like:

> When updating your cluster, the cluster version operator (CVO) creates a `status.history` entry to record the update.  This can be useful to decide if history-related bugs apply to your cluster.  It can also be used to correlate changes in cluster behavior with one potential trigger, although correlation alone does not imply causation.

File: modules/gathering-clusterversion-history.adoc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+====
+
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have installed the OpenShift CLI (`oc`).
+
+.Procedure
+
+. View a detailed summary of cluster specifications, update availability, and update history:
```
Comment (wking) [inline modules/gathering-clusterversion-history.adoc:23]: nit: looking at `.status.history` just gives you "update history".  It would be `availableUpdates` for availability, and... something under `spec` for cluster specifications?  "detailed summary" also seems like it's pulling in two directions, and `history` seems more on the detailed side of things and less on the summary side of things.  Maybe something like:

> View cluster update history:

?

File: modules/gathering-clusterversion-history.adoc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
++
+[source,terminal]
+----
+$ oc get clusterversion version -o json | jq .status.history
```
Comment (wking) [inline modules/gathering-clusterversion-history.adoc:27]: Dropping the `jq` exposure (you don't list it as a prereq, and reducing the number of prereqs is good) with [JSONPath][1]:

```console
$ oc get -o 'jsonpath={range .status.history[*]}{@}{"\n"}{end}' clusterversion version
{"completionTime":"2023-11-07T20:26:04Z","image":"quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4","startedTime":"2023-11-07T19:11:36Z","state":"Completed","verified":true,"version":"4.13.19"}
{"completionTime":"2023-10-04T18:53:29Z","image":"quay.io/openshift-release-dev/ocp-release@sha256:eac141144d2ecd6cf27d24efe9209358ba516da22becc5f0abc199d25a9cfcec","startedTime":"2023-10-04T17:26:31Z","state":"Completed","verified":true,"version":"4.13.13"}
{"completionTime":"2023-09-26T14:21:43Z","image":"quay.io/openshift-release-dev/ocp-release@sha256:371328736411972e9640a9b24a07be0af16880863e1c1ab8b013f9984b4ef727","startedTime":"2023-09-26T14:02:33Z","state":"Completed","verified":false,"version":"4.13.12"
```

That's not quite as pretty as the newlined output that `jq` produces, but it's not completely illegible.  Another alternative with more opinions about how to format the properties in each entry:

```console
$ oc get -o 'jsonpath={range .status.history[*]}{.startedTime}{" "}{.completionTime}{" "}{.state}{" "}{.version}{" (verified: "}{.verified}{", "}{.image}{") "}{.acceptedRisks}{"\n"}{end}' clusterversion version
2023-11-07T19:11:36Z 2023-11-07T20:26:04Z Completed 4.13.19 (verified: true, quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4) 
2023-10-04T17:26:31Z 2023-10-04T18:53:29Z Completed 4.13.13 (verified: true, quay.io/openshift-release-dev/ocp-release@sha256:eac141144d2ecd6cf27d24efe9209358ba516da22becc5f0abc199d25a9cfcec) 
2023-09-26T14:02:33Z 2023-09-26T14:21:43Z Completed 4.13.12 (verified: false, quay.io/openshift-release-dev/ocp-release@sha256:371328736411972e9640a9b24a07be0af16880863e1c1ab8b013f9984b4ef727)
```

There's also:

```console
$ oc describe clusterversions/version
...
  Desired:
    Channels:
      candidate-4.13
      candidate-4.14
      fast-4.13
      fast-4.14
      stable-4.13
    Image:    quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4
    URL:      https://access.redhat.com/errata/RHSA-2023:6130
    Version:  4.13.19
  History:
    Completion Time:    2023-11-07T20:26:04Z
    Image:              quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4
    Started Time:       2023-11-07T19:11:36Z
    State:              Completed
    Verified:           true
    Version:            4.13.19
    Completion Time:    2023-10-04T18:53:29Z
    Image:              quay.io/openshift-release-dev/ocp-release@sha256:eac141144d2ecd6cf27d24efe9209358ba516da22becc5f0abc199d25a9cfcec
    Started Time:       2023-10-04T17:26:31Z
    State:              Completed
    Verified:           true
    Version:            4.13.13
    Completion Time:    2023-09-26T14:21:43Z
    Image:              quay.io/openshift-release-dev/ocp-release@sha256:371328736411972e9640a9b24a07be0af16880863e1c1ab8b013f9984b4ef727
    Started Time:       2023-09-26T14:02:33Z
    State:              Completed
    Verified:           false
    Version:            4.13.12
  Observed Generation:  4
  Version Hash:         CMLl3sLq-EA=
Events:                 <none>
```

if we think that pretty-formatting of the `history` part is worth the hassle of scrolling through the non-history part of the structure.

And if we expect this to be something that folks need with regularity, in isolation from other parts of ClusterVersion, we can add new `oc ...` subcommands to make it easy to get at this information without needing `jq` or JSONPath.

[1]: https://kubernetes.io/docs/reference/kubectl/jsonpath/

Comment (wking) [reply modules/gathering-clusterversion-history.adoc:27]: Also, these options are not exclusive.  You could present both a JSONPath example and a `jq` example if you like.

**Review by LalatenduMohanty — APPROVED** [review]

**Review by LalatenduMohanty — APPROVED** [review]
/lgtm

File: modules/gathering-clusterversion-history.adoc [hunk 5]
```diff
@@ -0,0 +1,82 @@
+// Module included in the following assemblies:
+//
+// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="gathering-clusterversion-history_{context}"]
+= Gathering ClusterVersion history 
+
+When updating your cluster, the cluster version operator (CVO) creates a `status.history` entry to record the update. This can be useful to determine if history-related bugs apply to your cluster. The entry can also reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation.
```
Comment (abrennan89) [inline modules/gathering-clusterversion-history.adoc:9]:
```suggestion
When updating your cluster, the cluster version Operator (CVO) creates a `status.history` entry to record the update. This can be useful to determine if history-related bugs apply to your cluster. The entry can also reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
```
"Operator" should be capitalized - I also wonder, should it be "Cluster Version Operator" if that's an official name?

Suggestion: It may be worth adding an attribute for the Operator name if one doesn't already exist.

Comment (abrennan89) [reply modules/gathering-clusterversion-history.adoc:9]: Would it be helpful to users to mention here where the "`status.history` entry" is created? Is this part of some system logs, etc?

File: modules/gathering-clusterversion-history.adoc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+= Gathering ClusterVersion history 
+
+When updating your cluster, the cluster version operator (CVO) creates a `status.history` entry to record the update. This can be useful to determine if history-related bugs apply to your cluster. The entry can also reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
+
+[NOTE]
+====
+The ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the least relevant entries in the history are pruned to accommodate the limit. 
+====
+
+It is possible to view the ClusterVersion history though the web console and through the CLI.
```
Comment (abrennan89) [inline modules/gathering-clusterversion-history.adoc:16]:
```suggestion
It is possible to view the ClusterVersion history in the {product-title} web console, or by using the {oc-first}.
```
Per the IBMSG, do not use "through" to mean "by using":
image

https://www.ibm.com/docs/en/ibm-style?topic=word-usage#through

Comment (abrennan89) [reply modules/gathering-clusterversion-history.adoc:16]: Also suggest using the attributes here for the web console and CLI names

File: modules/gathering-clusterversion-history.adoc [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+. From the web console, click *Administration* -> *Cluster Settings* and review the contents of the *Details* tab.
+
+To view the ClusterVersion history through the CLI:
+
+.Prerequisites
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have installed the OpenShift CLI (`oc`).
+
+.Procedure
+
```
Comment (abrennan89) [inline modules/gathering-clusterversion-history.adoc:35]: There should only be one list of Prerequisites and one Procedure per module. This should probably be split into 2 modules, or reformatted so that the steps are maybe optional steps in a single procedure.

File: modules/gathering-clusterversion-history.adoc [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+. From the web console, click *Administration* -> *Cluster Settings* and review the contents of the *Details* tab.
+
+To view the ClusterVersion history through the CLI:
+
+.Prerequisites
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have installed the OpenShift CLI (`oc`).
```
Comment (abrennan89) [inline modules/gathering-clusterversion-history.adoc:32]:
```suggestion
* You have installed the {oc-first}.
```

File: modules/gathering-clusterversion-history.adoc [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+.Prerequisites
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have access to the {product-title} web console.
+
+.Procedure
+
+. From the web console, click *Administration* -> *Cluster Settings* and review the contents of the *Details* tab.
+
+To view the ClusterVersion history through the CLI:
```
Comment (abrennan89) [inline modules/gathering-clusterversion-history.adoc:28]:
```suggestion
To view the ClusterVersion history through the {oc-first}:
```

File: modules/gathering-clusterversion-history.adoc [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+.Prerequisites
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have installed the OpenShift CLI (`oc`).
+
+.Procedure
+
+. View cluster update history:
```
Comment (abrennan89) [inline modules/gathering-clusterversion-history.adoc:36]: If it's just one step in a procedure, use a bullet instead of a numbered list: 
image

https://redhat-documentation.github.io/modular-docs/#con-creating-procedure-modules_writing-mod-docs

File: modules/gathering-clusterversion-history.adoc [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+[source,terminal]
+----
+$ oc describe clusterversions/version
+----
++
+.Example output
+[source,terminal]
+----
+  $ oc describe clusterversions/version
+...
```
Comment (abrennan89) [inline modules/gathering-clusterversion-history.adoc:47]: Do not include the command in the example output code block:

image

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

**Review by abrennan89 — COMMENTED** [review]
Left some comments @obrown1205 

It looks like this may need to be split into 2 procedure modules, since we are only supposed to include one set of prereqs and one procedure per module.

File: modules/gathering-clusterversion-history.adoc (hunk 5 above)
Comment (obrown1205) [reply modules/gathering-clusterversion-history.adoc:9]: good call - I'll confirm with the engineers :)

Comment (obrown1205) [reply modules/gathering-clusterversion-history.adoc]: I've removed that sentence since it was more confusing than accurate. :) thanks!

File: modules/gathering-clusterversion-history.adoc [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="gathering-clusterversion-history_{context}"]
+= Gathering ClusterVersion history 
+
+The cluster version Operator (CVO) records updates made to a cluster, known as the ClusterVersion history. The entries can reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
+
+[NOTE]
+====
+The ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the least relevant entries in the history are pruned to accommodate the limit.
```
Comment (JianLi-RH) [inline modules/gathering-clusterversion-history.adoc:13]: hi @obrown1205 Shall we put more details about the limitation in the doc? like below in related bug, at least put the limit size 100 here:
```
Initial version is kept
all minor version updates are kept
the latest z streams updates are kept
however, the oldest z stream updates (in previous minor versions), which are not considered as important to the overall history of the cluster, are deleted from the history
```

**Comment by JianLi-RH** [comment]
cc @dis016 for him to better understand the CVO history

File: modules/gathering-clusterversion-history.adoc (hunk 12 above)
Comment (obrown1205) [reply modules/gathering-clusterversion-history.adoc:13]: can do :)

Comment (obrown1205) [reply modules/gathering-clusterversion-history.adoc]: @LalatenduMohanty what do you think of this?

The initial, minor, and z-stream version updates are stored by the ClusterVersion history. However, the ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the oldest z stream updates in previous minor versions are pruned to accommodate the limit.

Comment (LalatenduMohanty) [reply modules/gathering-clusterversion-history.adoc]: This looks mostly correct . But I do not think we need to go for writing the details. Anyone who will go through history should know that we might have modified the history if the number of entries are 100.

**Review by LalatenduMohanty — APPROVED** [review]
/lgtm

**Comment by obrown1205** [comment]
@abrennan89 could you take another look? :) I appreciate your feedback!

File: modules/gathering-clusterversion-history.adoc [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// Module included in the following assemblies:
+//
+// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="gathering-clusterversion-history_{context}"]
+= Gathering ClusterVersion history 
+
+The cluster version Operator (CVO) records updates made to a cluster, known as the ClusterVersion history. The entries can reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation.
```
Comment (kcarmichael08) [inline modules/gathering-clusterversion-history.adoc:9]:
```suggestion
The Cluster Version Operator (CVO) records updates made to a cluster, known as the ClusterVersion history. The entries can reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
```

Comment (kcarmichael08) [reply modules/gathering-clusterversion-history.adoc:9]: Capitalize Operator names. See https://redhat-documentation.github.io/supplementary-style-guide/#operator

File: modules/gathering-clusterversion-history.adoc [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="gathering-clusterversion-history_{context}"]
+= Gathering ClusterVersion history 
+
+The cluster version Operator (CVO) records updates made to a cluster, known as the ClusterVersion history. The entries can reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
+
+[NOTE]
+====
+The initial, minor, and z-stream version updates are stored by the ClusterVersion history. However, the ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the oldest z stream updates in previous minor versions are pruned to accommodate the limit.
```
Comment (kcarmichael08) [inline modules/gathering-clusterversion-history.adoc:13]: suggestion:

```suggestion
The initial, minor, and z-stream version updates are stored by the ClusterVersion history. However, the ClusterVersion history has a size limit. If the limit is reached, the oldest z-stream updates in previous minor versions are pruned to accommodate the limit. 
```

File: modules/gathering-clusterversion-history.adoc [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+= Gathering ClusterVersion history 
+
+The cluster version Operator (CVO) records updates made to a cluster, known as the ClusterVersion history. The entries can reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
+
+[NOTE]
+====
+The initial, minor, and z-stream version updates are stored by the ClusterVersion history. However, the ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the oldest z stream updates in previous minor versions are pruned to accommodate the limit. 
+====
+
+It is possible to view the ClusterVersion history by using the {product-title} web console and using the {oc-first}.
```
Comment (kcarmichael08) [inline modules/gathering-clusterversion-history.adoc:16]:
```suggestion
You can view the ClusterVersion history by using the {product-title} web console or by using the {oc-first}.
```

File: modules/gathering-clusterversion-history-cli.adoc [hunk 16]
```diff
@@ -0,0 +1,62 @@
+// Module included in the following assemblies:
+//
+// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="gathering-clusterversion-history-cli_{context}"]
+= Gathering ClusterVersion history using the {oc-first}
+
+To view the ClusterVersion history using the {oc-first}:
```
Comment (kcarmichael08) [inline modules/gathering-clusterversion-history-cli.adoc:9]:
```suggestion
You can view the ClusterVersion history by using the {oc-first}.
```

Comment (kcarmichael08) [reply modules/gathering-clusterversion-history-cli.adoc:9]: Should be a sentence because the next item is a section. Also, see [IBM Style guidance](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#u) on "by using".

File: modules/gathering-clusterversion-history-cli.adoc [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+.Prerequisites
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have installed the {oc-first}.
+
+.Procedure
+
+. View cluster update history:
```
Comment (kcarmichael08) [inline modules/gathering-clusterversion-history-cli.adoc:17]:
```suggestion
. View the cluster update history by entering the following command:
```

Comment (kcarmichael08) [reply modules/gathering-clusterversion-history-cli.adoc:17]: See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command:

"If a step in a procedure is to run a command, make sure that the step text includes an explicit instruction to "run" or "enter" the command."

File: modules/gathering-clusterversion-history-console.adoc [hunk 18]
```diff
@@ -0,0 +1,17 @@
+// Module included in the following assemblies:
+//
+// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="gathering-clusterversion-history-console_{context}"]
+= Gathering ClusterVersion history in the {product-title} web console
+
+To view the ClusterVersion history in the {product-title} web console:
```
Comment (kcarmichael08) [inline modules/gathering-clusterversion-history-console.adoc:9]:
```suggestion
You can view the ClusterVersion history in the {product-title} web console.
```

**Review by kcarmichael08 — COMMENTED** [review]
Nice job! I marked a few style issues to fix, but overall looks great!
